src/client/datascience/data-viewing/dataViewer.ts

```diff
@@ -99,23 +114,31 @@ export class DataViewer extends WebviewPanelHost<IDataViewerMapping> implements
     protected onMessage(message: string, payload: any) {
         switch (message) {
             case DataViewerMessages.GetAllRowsRequest:
-                this.getAllRows().ignoreErrors();
+                this.getAllRows(payload as string).ignoreErrors();
                 break;
 
             case DataViewerMessages.GetRowsRequest:
                 this.getRowChunk(payload as IGetRowsRequest).ignoreErrors();
                 break;
 
+            case DataViewerMessages.GetSliceRequest:
+                this.getSlice(payload as IGetSliceRequest).ignoreErrors();
+                break;
+
             default:
                 break;
         }
 
         super.onMessage(message, payload);
     }
 
-    private getDataFrameInfo(): Promise<IDataFrameInfo> {
-        if (!this.dataFrameInfoPromise) {
-            this.dataFrameInfoPromise = this.dataProvider ? this.dataProvider.getDataFrameInfo() : Promise.resolve({});
+    private getDataFrameInfo(sliceExpression?: string): Promise<IDataFrameInfo> {
+        // If requesting a new slice, refresh our cached info promise
+        if (!this.dataFrameInfoPromise || sliceExpression !== this.currentSlice) {
```

Comment:
Might be reading this wrong. But this is the case where the slice expression has been updated. In that case would you want to cancel the existing dataFrameInfoPromise and start a new one? Looks like if there is already an existing promise here then it will just return the existing one. Doesn't seem to match up with how I read the comment.

Reply:
`this.currentSlice` is the cache of the last slice expression we evaluated (that might not have been clear from the name). If we're asking for a new slice expression that doesn't match the one we already have cached, then we ask for a new one. And of course if there isn't a cached promise already then we need to ask for one anyway. Did I understand your question correctly?

Reply:
Hmmm, maybe I'm reading it wrong. It was the ! on the dataFrameInfoPromise that I was not sure about. Is that ever cleared? Looks to me like currentSlice would never update here after that promise is set for the first time.

Reply:
The `!this.dataFrameInfoPromise` check is prexisting, i.e. if we don't have anything cached, we need to ask the data provider for the info. It's never cleared, but `this.currentSlice` and `this.dataFrameInfoPromise` are updated when we ask for a slice expression that is different than the one we have cached because the second condition in the if will evaluate to false even if the promise is set. 

Does this make any sense 😂